modules/xfem/src/userobjects/InterfaceMeshCut3DUserObject.C

```diff
@@ -0,0 +1,243 @@
+      }
+      normal[4 + i] = normal_at_edge;
+    }
+    _pseudo_normal.insert(std::make_pair(elem->id(), normal));
+  }
+}
+
+Point
+InterfaceMeshCut3DUserObject::nodeNormal(const unsigned int & node_id)
+{
+  Point normal(0.0);
+
+  for (const auto & node_neigh_elem_id : _node_to_elem_map[node_id])
+  {
+    const auto & elem = _cutter_mesh->elem_ref(node_neigh_elem_id);
+    Plane elem_plane(elem.node_ref(0), elem.node_ref(1), elem.node_ref(2));
+    normal += elem_plane.unit_normal(elem.node_ref(0));
+  }
+
+  unsigned int num = _node_to_elem_map[node_id].size();
+  return normal / num;
+}
+
+bool
+InterfaceMeshCut3DUserObject::cutElementByGeometry(const Elem * /*elem*/,
+                                                   std::vector<Xfem::CutEdge> & /*cut_edges*/,
+                                                   std::vector<Xfem::CutNode> & /*cut_nodes*/) const
+{
+  mooseError("invalid method for 3D mesh cutting");
+  return false;
+}
+
+bool
+InterfaceMeshCut3DUserObject::cutElementByGeometry(const Elem * elem,
+                                                   std::vector<Xfem::CutFace> & cut_faces) const
+{
+  bool elem_cut = false;
+
+  if (elem->dim() != 3)
+    mooseError("The structural mesh to be cut by a surface mesh must be 3D!");
+
+  for (unsigned int i = 0; i < elem->n_sides(); ++i)
+  {
+    // This returns the lowest-order type of side.
+    std::unique_ptr<const Elem> curr_side = elem->side_ptr(i);
+    if (curr_side->dim() != 2)
+      mooseError("In cutElementByGeometry dimension of side must be 2, but it is ",
+                 curr_side->dim());
+    unsigned int n_edges = curr_side->n_sides();
+
+    std::vector<unsigned int> cut_edges;
+    std::vector<Real> cut_pos;
+
+    for (unsigned int j = 0; j < n_edges; j++)
+    {
+      // This returns the lowest-order type of side.
+      std::unique_ptr<const Elem> curr_edge = curr_side->side_ptr(j);
+      if (curr_edge->type() != EDGE2)
+        mooseError("In cutElementByGeometry face edge must be EDGE2, but type is: ",
+                   libMesh::Utility::enum_to_string(curr_edge->type()),
+                   " base element type is: ",
+                   libMesh::Utility::enum_to_string(elem->type()));
+      const Node * node1 = curr_edge->node_ptr(0);
+      const Node * node2 = curr_edge->node_ptr(1);
+
+      for (const auto & cut_elem : _cutter_mesh->element_ptr_range())
+      {
+        std::vector<Point> vertices;
+
+        for (auto & node : cut_elem->node_ref_range())
+        {
+          Point & this_point = node;
+          vertices.push_back(this_point);
+        }
+
+        Point intersection;
+        if (Xfem::intersectWithEdge(*node1, *node2, vertices, intersection) &&
+            std::find(cut_edges.begin(), cut_edges.end(), j) == cut_edges.end())
+        {
+          cut_edges.push_back(j);
+          cut_pos.emplace_back(Xfem::getRelativePosition(*node1, *node2, intersection));
```

Comment:
Is there a reason `push_back` won't work here? It would be more consistent with the other code (like the line just above it), and my understanding is that `push_back` in general is safer.